[FLOC-3270] Add a new interface for IBlockDeviceAPIs with storage profiles.

sarum90 opened the pull request:
Introduce a new interface for drivers that are capable of creating profiles.

In the long term drivers can choose to provide IProfiledBlockDeviceAPI, and if they do, then flocker will call create_volume_with_profile if the user attempts to create a volume with a specified storage profile over the control API.

I suppose this might be arguably more of a design review disguised as a tiny code review.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/clusterhq/flocker/2066)
<!-- Reviewable:end -->

myechuri commented on flocker/node/agents/blockdevice.py:
```diff
@@ -944,6 +944,42 @@ def get_device_path(blockdevice_id):
+    An interface for ``IBlockDeviceAPI`` implementers that are capable of
+    creating volumes with a specific profile.
+    """
+
```
How about adding generation/version attribute to the interface?

generation = Attribute('1')

sarum90 replied:
If it's an interface version number do we want it as an attribute or in the name?

IProfiledBlockDeviceAPIV1 or something.

itamarst replied:
Version numbers as attributes of implementations would allow us to... change `IProfiledBlockDeviceAPI` within our code rather than copy/paste alter for new versions. But that means we don't have good way to refer to old version in any adapter code, so seems better to just copy/paste for this particular use case.

sarum90 replied:
Sticking as is for now. Will re-visit before we have people outside the company try to implement one of these.

myechuri commented on flocker/node/agents/test/test_blockdevice.py:
```diff
@@ -2281,6 +2281,45 @@ def setUp(self):
+        Verify that the profile for a volume created with one of the mandatory
+        profiles is respected.
+        """
+        for profile in [u'gold', u'silver', u'bronze']:
```
Should we define below in blockdevice.py?

class MandatoryProfiles(Values):
    """
    Mandatory Storage Profiles to be implemented by ``IProfiledBlockDeviceAPI``
    """
    GOLD = ValueConstant(u'gold')
    SILVER = ValueConstant(u'silver')
    BRONZE = ValueConstant(u'bronze')
    DEFAULT = ValueConstant(u'silver')

itamarst replied:
I suggest default belongs outside the class, but yeah having explicit constants is a good idea.

itamarst replied:
Or hm, maybe it does belong there. But perhaps `DEFAULT = SILVER`.

sarum90 replied:
Added.

itamarst commented on flocker/node/agents/blockdevice.py:
```diff
@@ -944,6 +944,42 @@ def get_device_path(blockdevice_id):
         """
 
 
+class IProfiledBlockDeviceAPI(IBlockDeviceAPI):
```
I don't think needs to inherit from `IBlockDeviceAPI`. It can just be an additional interface a class implements in addition to `IBlockDeviceAPI`, with additional methods.

itamarst commented:
Thanks, address and merge; I'm assuming a decent chance we'll end up tweaking these a bit once we start using them so we'll have at least one review with actual usage in mind.

sarum90 commented:
Addressed comments. I also removed 1 of the methods from the interface. (We can add it back later if we need it - I will implement that). If this code was in-use I might re-submit it for code review... given that it is not in use any decisions can be revisited and questioned in subsequent reviews, so I intend to submit once I get greens on buildbot.
